### PR TITLE
subscription iterators

### PR DESCRIPTION
**Pull request by aricart**
- Changed subscriptions to use iterators
- Made it possible to still specify a callback should it be necessary by setting the callback function as an option in the subscription
- See tests/iterators_test.ts for an example of what is possible

### CODE REVIEW

**Comment by aricart** [comment]
@kozlovic skip reviewing all the changes in the tests to accommodate this feature, except for tests/iterators_test.ts.

File: nats-base-client/types.ts [hunk 1]
```diff
@@ -117,14 +118,14 @@ export interface Req extends Base {
   token: string;
 }
 
-export interface Sub extends Base {
-  sid: number;
-  queue?: string | null;
-}
+// export interface Sub extends Base {
```
Comment (kozlovic) [inline nats-base-client/types.ts:121]: Should you remove?

File: nats-base-client/types.ts [hunk 2]
```diff
@@ -117,14 +118,14 @@ export interface Req extends Base {
...
+//   sid: number;
+//   queue?: string | null;
+// }
 
-export function defaultSub(): Sub {
-  return { sid: 0, subject: "", received: 0 } as Sub;
-}
+// export function defaultSub(): Sub {
```
Comment (kozlovic) [inline nats-base-client/types.ts:126]: same

File: tests/iterators_test.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
+import { connect } from "../src/connect.ts";
```
Comment (kozlovic) [inline tests/iterators_test.ts:1]: copyright banner missing

**Pull request merged**